### ISSUE TITLE
Support the DC upload

### ISSUE THREAD

**Issue by bberenberg** [Atlas-Authority/atlassian-marketplace-release-action#5]
Not sure if we properly support DC uploads or not, see https://community.developer.atlassian.com/t/releasing-server-data-center-apps-using-marketplace-rest-api/34286/3?u=george-soteri

**Comment by nhan-nguyen-se** [comment]
Currently the release action omits field `compatibilities` when create new version. As stated in the [API documentation ](https://developer.atlassian.com/platform/marketplace/rest/v2/api-group-app-versions/#api-addons-addonkey-versions-post), if that field is omitted then marketplace will use the previous `compatibilities` which is probably the latest DC version (because DC version number is greater than the server one). As the result, release action only create new version for DC.

Then I made the change in release action to include `compatibilities` which specifies both server and DC compatibility information, then I got the issue stated in the community link above: There is only one version is created (and it compatible with both server and DC). It does not like what happen if the app is release via web interface (two separated versions with the same version name but different build numbers.)

In the thread they said about adding custom header to request but I don't know how to use it. The header value is weird and I'm asking the author for more information.

**Comment by nhan-nguyen-se** [comment]
Tried to use header `X-Mpac-DataCenter-BuildNumber` to create single version name with different build numbers for server and DC for example app but got below error:

> {"code":400,"message":{"errors":[{"message":"Sorry! You can no longer publish a new Data Center app with server compatibility. Learn more: https://community.developer.atlassian.com/t/end-of-new-server-sales-resources-and-support-for-the-ecosystem/42325."}]}}

As mentioned in the link, newly submitted apps are not allowed to publish DC version with server compatibility. Also tried on web interface (I created a test marketplace account and a vender) and it does not allow to do that too.

So our example app does not have the same state with our prod apps making it inappropriate for testing purpose.

**Comment by nhan-nguyen-se** [comment]
I'm still waiting for a response from community on how to use `X-Mpac-DataCenter-BuildNumber` header. My guess is that we need to explicitly provide the server build number in versions create request body, and also include that header with a value of server build number increased by 10.

Once we know how to use that header, we need to figure out a way to test it since our example app is useless for marketplace place upload testing.

**Comment by bberenberg** [comment]
I am confused with what you're saying. We should be able to upload new versions of apps. Not brand new apps in their entirety.

**Comment by nhan-nguyen-se** [comment]
Yes. For the existing public apps (like our production apps) we can still publish their DC version with with server compatibility using web interface (same version name but different build number). The problem is how to archive the same thing by using REST API. The community suggest adding a header in the request and I'm asking them for more information.

When I got more information about how to use the API from the community I need to test it on a dummy app. The problem is I cannot use our example app for that purpose because Atlassian does not allow to publish its DC version with with server compatibility. More information here https://community.developer.atlassian.com/t/end-of-new-server-sales-resources-and-support-for-the-ecosystem/42325

**Comment by nhan-nguyen-se** [comment]
And this reply from Atlassian staff https://community.developer.atlassian.com/t/reminder-we-will-stop-accepting-new-server-app-submissions-from-may-1-exceptions-to-be-made-for-bamboo-and-crowd/47042/3

**Comment by bberenberg** [comment]
This doesn't make sense Nhan. The thread you're asking how to get the server version number, that is dynamically generated by incrementing the previous version and leaving a consistent gap. You can look at old NAFJ versions to see what the gap is. 

At the same time, you're talking about publishing DC version being blocked. This is a limitation yes, but this seems like something we need to request Atlassian to unlock on our test listing right? Not anything to do with the actual upload functionality we have here.

**Comment by bberenberg** [comment]
We don't need to publish public versions of server and dc, we need private versions. 

The server build number is the previous build number incremented by a set interval as I described above. Please go look at past numbers for NAFJ versions to see this.

**Comment by nhan-nguyen-se** [comment]
Okay. I got what you mean about server build number.

Even for private versions, we still need to ask Atlassian to unlock our testing app ([example-confluence-server-app](https://github.com/Atlas-Authority/example-confluence-server-app)) so that we can publish DC version with server compatibility (which mean same version name but different build number). Currently we can only publish DC version, not DC version with server compatibility.

I mentioned _public_ because I see latest versions of our app are public:
- https://marketplace.atlassian.com/rest/2/addons/com.riadalabs.jira.plugins.notificationassistant/versions/latest?hosting=dataCenter
- https://marketplace.atlassian.com/rest/2/addons/com.riadalabs.jira.plugins.notificationassistant/versions/latest?hosting=server

**Comment by bberenberg** [comment]
But we previously had server and dc at https://marketplace.atlassian.com/manage/apps/1227599/versions and you don't need approval from them to publish a private version, only to make it public.

**Comment by nhan-nguyen-se** [comment]
I don't have access to the link because it's private. But I guess server and DC have the same version and the same build number? If so that is the issue stated in the community thread.

**Comment by bberenberg** [comment]
Here is the example from NAFJ:
<img width="496" alt="Google Chrome 2022-04-14 at 11 06 07@2x" src="https://user-images.githubusercontent.com/4259775/163429439-8dcda166-06db-485e-9767-ea892e0c2835.png">

They're not the same, they are incremented.

**Comment by nhan-nguyen-se** [comment]
No I'm talking about our testing app: [example-confluence-server-app](https://github.com/Atlas-Authority/example-confluence-server-app). The link you mention earlier is our testing app.

As I said before, example-confluence-server-app has problem, not our prod apps.

**Comment by bberenberg** [comment]
You're linking a Git Repo, this doesn't help as you have created multiple apps on marketplace. Which app are you using there? Is it https://marketplace.atlassian.com/manage/apps/1227599/versions ?

**Comment by nhan-nguyen-se** [comment]
Yes https://marketplace.atlassian.com/manage/apps/1227599/versions

**Comment by nhan-nguyen-se** [comment]
@bberenberg So are you working with Atlassian to unlock https://marketplace.atlassian.com/manage/apps/1227599/versions?

**Comment by bberenberg** [comment]
Yes I am

**Comment by boris-moduscreate** [comment]
Hey so I got Atlassian to start looking at this, but we're still not sure what you need here. https://marketplace.atlassian.com/manage/apps/1227599/versions is showing both DC and Server issues in the history. We only want this action to push private versions. What exactly do we need from Atlassian here?

**Comment by nhan-nguyen-se** [comment]
@bberenberg I don't have access to https://marketplace.atlassian.com/manage/apps/1227599/versions. Can you screenshot it to me?

**Comment by boris-moduscreate** [comment]
You should be able to access it as the devops account?

**Comment by nhan-nguyen-se** [comment]
@boris-moduscreate We need the ability to create the **_same_** version for both DC and server with **_different_** build number, like NAFJC:
<img width="534" alt="image" src="https://user-images.githubusercontent.com/39516534/184797897-bafe9c49-229b-4144-ae43-03cd427bfd4c.png">

We unable to do the same with our example app:
<img width="593" alt="image" src="https://user-images.githubusercontent.com/39516534/184797993-d6c161e6-a660-4fd1-bab0-a22fd2a384ba.png">
Here DC and server have the same version and the same build number.

The [community](https://community.developer.atlassian.com/t/releasing-server-data-center-apps-using-marketplace-rest-api/34286/3?u=george-soteri) suggest to use `X-Mpac-DataCenter-BuildNumber` header, I tried and got this error:
`{"code":400,"message":{"errors":[{"message":"Sorry! You can no longer publish a new Data Center app with server compatibility. Learn more: https://community.developer.atlassian.com/t/end-of-new-server-sales-resources-and-support-for-the-ecosystem/42325."}]}}`

So we need Atlassian Support on how to use that header for our example app. (Using that header is confirmed a solution by community so I think the issue is about our example app)

**Comment by boris-moduscreate** [comment]
@nhan-nguyen-se would it make sense to update this to use: https://github.com/marvelution/mpac-version-publisher/ ?

**Comment by nhan-nguyen-se** [comment]
Yeah we can give it a try.

If I recall correctly, this is stuck because we can't test this with our example app and waiting for Atlassian to unlock our example app. Atlassian adds some restrictions to newly created app, so there are thing we can do with old apps (our prod apps), but can't do with example app. An example is ability to public the same version with different build numbers for server and DC. It is possible with old app, but not with new app. So either using the new tool `mpac-version-publisher` or our GHA, we probably have to test it with our prod app.

**Comment by boris-moduscreate** [comment]
Ok so I looked at https://marketplace.atlassian.com/manage/apps/1227599/versions and we have both server and dc versions in place, so the listing "works". The goal should be identical versions, but different build numbers. Not sure how much work it would be to reproduce our old state and confirm the cause.
